src/java.security.jgss/share/classes/sun/security/krb5/SCDynamicStoreConfig.java

```diff
@@ -102,45 +65,64 @@ public Boolean run() {
-        Hashtable<String, Object> stanzaTable = getKerberosConfig();
-        if (stanzaTable == null) {
+        List<String> list = getKerberosConfig();
+        if (list == null) {
             throw new IOException(
                     "Could not load configuration from SCDynamicStore");
         }
-        if (DEBUG) System.out.println("Raw map from JNI: " + stanzaTable);
-        return convertNativeConfig(stanzaTable);
-    }
+        if (DEBUG) System.out.println("Raw map from JNI: " + list);
 
-    @SuppressWarnings("unchecked")
-    private static Hashtable<String, Object> convertNativeConfig(
-            Hashtable<String, Object> stanzaTable) throws IOException {
-        // convert SCDynamicStore realm structure to Java realm structure
-        Hashtable<String, ?> realms =
-                (Hashtable<String, ?>) stanzaTable.get("realms");
-        if (realms == null || realms.isEmpty()) {
-            throw new IOException(
-                    "SCDynamicStore contains an empty Kerberos setting");
-        }
-        stanzaTable.remove("realms");
-        Hashtable<String, Object> realmsTable = convertRealmConfigs(realms);
-        stanzaTable.put("realms", realmsTable);
-        WrapAllStringInVector(stanzaTable);
-        if (DEBUG) System.out.println("stanzaTable : " + stanzaTable);
-        return stanzaTable;
-    }
+        Hashtable<String,Object> v = new Hashtable<>();
+        Hashtable<String,Object> realms = new Hashtable<>();
+        Iterator<String> iterator = list.iterator();
+        String defaultRealm = null;
 
-    @SuppressWarnings("unchecked")
-    private static void WrapAllStringInVector(
-            Hashtable<String, Object> stanzaTable) {
-        for (String s: stanzaTable.keySet()) {
-            Object v = stanzaTable.get(s);
-            if (v instanceof Hashtable) {
-                WrapAllStringInVector((Hashtable<String,Object>)v);
-            } else if (v instanceof String) {
-                Vector<String> vec = new Vector<>();
-                vec.add((String)v);
-                stanzaTable.put(s, vec);
+        while (true) {
+            String nextRealm = iterator.next();
+            if (nextRealm == null) {
+                break;
+            }
+            if (defaultRealm == null) {
+                defaultRealm = nextRealm;
+                Hashtable<String,Object> dr = new Hashtable<>();
+                dr.put("default_realm", v1(defaultRealm));
+                v.put("libdefaults", dr);
+            }
+            Hashtable<String,Object> ri = new Hashtable<>();
+            Vector<String> kdcs = new Vector<>();
+            while (true) {
+                String nextKdc = iterator.next();
+                if (nextKdc == null) {
+                    break;
+                }
+                kdcs.add(nextKdc);
             }
+            if (!kdcs.isEmpty()) {
+                ri.put("kdc", kdcs);
+            }
+            realms.put(nextRealm, ri);
```

Comment:
Do you need to check if ri is empty?

Reply:
It should make no difference to applications whether there is an empty section or not, but I can remove it.